File: GFG POTD/011-October-2024/025-alternative-sorting.java
// https://www.geeksforgeeks.org/problems/alternative-sorting1311/1




//{ Driver Code Starts
// Initial Template for Java
import java.io.*;
import java.lang.*;
import java.util.*;

class Main {
    public static void main(String args[]) throws IOException {
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());

        while (t-- > 0) {

            ArrayList<Integer> array1 = new ArrayList<Integer>();
            String line = read.readLine();
            String[] tokens = line.split(" ");
            for (String token : tokens) {
                array1.add(Integer.parseInt(token));
            }
            ArrayList<Integer> v = new ArrayList<Integer>();
            int[] arr = new int[array1.size()];
            int idx = 0;
            for (int i : array1) arr[idx++] = i;

            v = new Solution().alternateSort(arr);

            for (int i = 0; i < v.size(); i++) System.out.print(v.get(i) + " ");

            System.out.println();
            System.out.println("~");
        }
    }
}
// } Driver Code Ends

// User function Template for Java
class Solution {
    // TC : O(N*log(N))
    // SC : O(N)
    public static ArrayList<Integer> alternateSort(int[] arr) {
        int size = arr.length;
        ArrayList<Integer> res = new ArrayList<>();
        Arrays.sort(arr);
        int i = size - 1;
        int j = 0;
        boolean flag = false;
        while (j <= i) {
            if (flag) {
                res.add(arr[j]);
                j ++;
            } else {
                res.add(arr[i]);
                i --;
            }
            flag = !flag;
        }
        return res;
    }
}

File: GFG POTD/011-October-2024/025-alternative-sorting.cpp
// https://www.geeksforgeeks.org/problems/alternative-sorting1311/1




//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;
// } Driver Code Ends

class Solution {
  private:
    // TC : O(N*log(N))
    // SC : O(N)
    vector<int> approach_1(vector<int>& arr) {
        int n = arr.size();
        vector<int> vec = arr;
        sort(vec.rbegin(), vec.rend());
        int i = 0, j = n - 1, idx1 = 0, idx2 = 1;
        while (idx1 < n) {
            arr[idx1] = vec[i];
            arr[idx2] = vec[j];
            i ++, j --;
            idx1 += 2, idx2 += 2;
        }
        return arr;
    }
    
    // TC : O(N*log(N)) Inserting N elements into set
    // SC : O(K)        Where K = No. of Unique elements
    vector<int> approach_2(vector<int>& arr) {
        int n = arr.size();
        set<int> st;
        for (int &num : arr) {
            st.insert(num);
        }
        
        int idx = 0;
        for (auto it = st.rbegin(); it != st.rend(); it ++) {
            arr[idx] = *it;
            idx += 2;
            if (idx >= n) break;
        }
        
        idx = 1;
        for (int it : st) {
            arr[idx] = it;
            idx += 2;
            if (idx >= n) break;
        }
        return arr;
    }
  public:
    vector<int> alternateSort(vector<int>& arr) {
        // return approach_1(arr);
        
        return approach_2(arr);
    }
};


//{ Driver Code Starts.
int main() {
    string ts;
    getline(cin, ts);
    int t = stoi(ts);
    while (t--) {
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        Solution obj;
        vector<int> ans = obj.alternateSort(arr);
        for (int i = 0; i < ans.size(); i++) {
            cout << ans[i] << " ";
        }
        cout << endl;
        cout << "~" << endl;
    }
    return 0;
}
// } Driver Code Ends